File: frontend/src/api/support.ts
import { z } from "zod";

import { apiClient } from "@/api/client";
import { isTelegramEnvironment } from "@/lib/telegram";

export const supportCategories = ["bug", "question", "idea", "other"] as const;
export const supportStatuses = ["waiting_support", "waiting_user", "resolved", "closed"] as const;

const messageSchema = z.object({
  id: z.string().uuid(),
  author_type: z.enum(["user", "admin", "system"]),
  body: z.string(),
  delivery_channel: z.enum(["in_app", "telegram"]),
  delivery_status: z.enum(["pending", "sent", "failed", "not_requested", "unavailable"]),
  created_at: z.string().datetime({ offset: true }),
});

const summarySchema = z.object({
  id: z.string().uuid(),
  category: z.enum(supportCategories),
  status: z.enum(supportStatuses),
  subject: z.string(),
  last_message_preview: z.string(),
  unread: z.boolean(),
  last_message_at: z.string().datetime({ offset: true }),
  created_at: z.string().datetime({ offset: true }),
});

const detailSchema = summarySchema.extend({
  source_page: z.string().nullable(),
  client: z.string(),
  app_version: z.string().nullable(),
  messages: z.array(messageSchema),
});

const listSchema = z.object({ items: z.array(summarySchema), total: z.number().int().nonnegative() });

export type SupportCategory = (typeof supportCategories)[number];
export type SupportStatus = (typeof supportStatuses)[number];
export type SupportMessage = z.infer<typeof messageSchema>;
export type SupportTicketSummary = z.infer<typeof summarySchema>;
export type SupportTicketDetail = z.infer<typeof detailSchema>;

export async function listSupportTickets(): Promise<SupportTicketSummary[]> {
  const { data } = await apiClient.get("/support/tickets");
  return listSchema.parse(data).items;
}

export async function createSupportTicket(input: {
  category: SupportCategory;
  message: string;
  page: string;
}): Promise<SupportTicketSummary> {
  const idempotencyKey = crypto.randomUUID();
  const { data } = await apiClient.post("/support/tickets", {
    category: input.category,
    message: input.message,
    page: input.page,
    client: isTelegramEnvironment() ? "telegram" : "browser",
    app_version: __FITNESS_BUILD_ID__,
    idempotency_key: idempotencyKey,
  });
  return summarySchema.parse(data);
}

export async function getSupportTicket(ticketId: string): Promise<SupportTicketDetail> {
  const { data } = await apiClient.get(`/support/tickets/${ticketId}`);
  return detailSchema.parse(data);
}

export async function sendSupportMessage(ticketId: string, message: string): Promise<SupportMessage> {
  const { data } = await apiClient.post(`/support/tickets/${ticketId}/messages`, {
    message,
    idempotency_key: crypto.randomUUID(),
  });
  return messageSchema.parse(data);
}

export async function closeSupportTicket(ticketId: string): Promise<void> {
  await apiClient.post(`/support/tickets/${ticketId}/close`);
}

File: backend/app/schemas/support.py
"""User and administrator contracts for in-app support."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

SupportCategory = Literal["bug", "question", "idea", "other"]
SupportStatus = Literal["waiting_support", "waiting_user", "resolved", "closed"]


class SupportTicketCreate(BaseModel):
    category: SupportCategory
    message: str = Field(min_length=3, max_length=3500)
    page: str = Field(default="", max_length=300)
    client: Literal["telegram", "browser"] = "browser"
    app_version: str = Field(default="", max_length=80)
    idempotency_key: uuid.UUID


class SupportMessageCreate(BaseModel):
    message: str = Field(min_length=1, max_length=3500)
    idempotency_key: uuid.UUID


class SupportMessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    author_type: Literal["user", "admin", "system"]
    body: str
    delivery_channel: Literal["in_app", "telegram"]
    delivery_status: Literal["pending", "sent", "failed", "not_requested", "unavailable"]
    created_at: datetime


class SupportTicketSummary(BaseModel):
    id: uuid.UUID
    category: SupportCategory
    status: SupportStatus
    subject: str
    last_message_preview: str
    unread: bool
    last_message_at: datetime
    created_at: datetime


class SupportTicketListResponse(BaseModel):
    items: list[SupportTicketSummary]
    total: int


class SupportTicketDetail(SupportTicketSummary):
    source_page: str | None = None
    client: str
    app_version: str | None = None
    messages: list[SupportMessageResponse]


class AdminSupportTicketSummary(SupportTicketSummary):
    user_id: uuid.UUID
    user_label: str


class AdminSupportTicketListResponse(BaseModel):
    items: list[AdminSupportTicketSummary]
    total: int
    page: int
    page_size: int
    waiting_support: int


class AdminSupportTicketDetail(SupportTicketDetail):
    user_id: uuid.UUID
    user_label: str


class SupportStatusUpdate(BaseModel):
    status: Literal["waiting_support", "waiting_user", "resolved", "closed"]
